man/im_extract.3
.TH IM_EXTRACT 3 "11 April 1990"
.SH NAME
im_extract_areabands, im_extract_bands, im_extract_area \- extract a portion of an image
.SH SYNOPSIS
.B #include <vips/vips.h>

int im_extract_areabands( IMAGE *in, IMAGE *out, 
.br
  int left, int top, int width, int height, int band, int nbands )

int im_extract_area( IMAGE *in, IMAGE *out,
.br
  int left, int top, int width, int height )

int im_extract_bands( IMAGE *in, IMAGE *out, 
.br
  int chsel, int nbands )

.SH DESCRIPTION
.B im_extract_areabands(3) 
extracts the rectangular portion of the image defined by 
.B left, 
.B top, 
.B width,
and
.B height of image 
.B in 
and writes the result to image 
.B out. 
The area must lie entirely within in the image. Selects the set of 
.B nbands
bands 
starting at band number
.B band
(numbering bands from zero).

Works for any size image, any number of bands, any type. Works for LABPACK
coded images too! But disallows band extraction in this case.

.B im_extract_area(3) 
is a convenience function which extracts an area from an
image, leaving the bands the same.

.B im_extract_bands(3) 
takes 
.B nbands
out of an image, starting from band
.B band.
So, for example, nbands == 2, bands == 1 will form a two band image from an RGB
image, where the two bands are the G and the B bands.

.SH RETURN VALUE
The function returns 0 on success and -1 on error.
.SH SEE ALSO
im_insert(3), im_lrjoin(3), im_lrmerge(3), im_stats(3),
im_region_region(3).
.SH COPYRIGHT
J. Cupitt,
.SH AUTHOR
J. Cupitt \- 11/04/1990


NEWS
VIPS changed from 7.10 to 7.12 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- IMAGE->Hist is deprecated, use im_history_get() instead
- im_append_Hist() removed
- IMAGE->Bbits is deprecated (now ignored)
- im_region_local() replaced by im_region_buffer()

VIPS enhancements
=================

- new backwards and forwards compatible VIPS file format: it has a new 
  metadata system that efficiently shares and propogates ICC profiles,
  EXIF data, etc. & whatever
- speed ups: new pixel buffer cache saves and reuses computations, uses liboil 
  where possible for a small speedup
- SMP scaling: simplified lock system improves SMP scaling, double-buffered 
  image writes overlap computation 
- lower mem use: new mmap window cache shares file mappings, history buffers 
  share and reuse image history metadata
- built-in support for new image formats: OpenEXR, Analyze7, CSV
- RGB16 and GREY16 image hints
- uses GOption for much better command-line argument handling
- better C++ binding: namespaces, VError derives from std::exception, image
  and number vector constants
- python binding 
- gcc sentinel attributes added
- added GValue as an arg type
- added im_benchmark() and friends for testing
- new functions: 
	im_analyze2vips(), im_benchmark*(), im_buildlut(),
	im_concurrency_get/set(), im_contrast_surface(),
	im_contrast_surface_raw(), im_copy_from(), im_copy_morph(),
	im_cp_desc_array(), im_cp_descv(), im_extract_areabands(),
	im_flood_blob_copy(), im_get_option_group(), im_grid(),
	im_header_exists(), im_header_map(), im_history_get(),
	im_invalidate(), im_isscalar(), im_lineset(), im_linreg(),
	im_meta*(), im_msb(), im_norm_dmask(), im_project(),
	im_raw2vips(), IM_REGION_ADDR_TOPLEFT(), im_tile_cache(),
	im_video_test()


VIPS changed from 7.8 to 7.10 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- no longer generates vips-config script, instead it makes .pc files for
  pkg-config ... replace any "vips-config" with "pkg-config vips-7.10"
- origin hint semantics changed ... it now records the position in the output
  image of the origin of the input image
- all float to int casts on pixel values are now floor() rather than rint(),
  for consistency with C casting semantics

VIPS enhancements
=================

- nip reworked for gtk+-2.4, now called nip2
- now based on glib, so VIPS no longer has it's own thread, plugin and data
  structure libraries
- i18n support, although not quite complete yet
- new error message API to make i18n possible (compatibility macros mean the
  old API still works)
- new 'start up VIPS' call, im_init_world(), sets up i18n and threads
  (optional to call it, but i18n won't fully work unless you do)
- C programs now just need "#include <vips/vips.h>", no other includes
  required
- wraps the Pango font rendering library, see im_text()
- new function flag: NOCACHE ... means the result of the call should not be
  cached, useful for video grab functions and paintbox functions
- im_embed() now supports a wider range of embedding styles, including a fast
  edge stretch
- all area operations use the new embed style to expand their input rather 
  than their output, so there are no more black borders on filtered images
- other new functions: im_render(), im_cache(), im_extract_bands(), 
  im_copy_swap(), im_rint(), im_make_xy(), im_init_world(), im_error(),
  im_warn(), im_diag(), im_rank_image(), im_replicate()
- many fixes and improvements to old funcs
- configure and portability improvements

Bug fixes
=========

- all serious bug fixes got back-ported to 7.8, I think


VIPS changes from 7.6 to 7.8 (again, not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- output format options specified in filenames: new prototypes for load and
  save tiff and jpeg
- C++ API name changes ... error => VError
- include path change
    - C includes <vips/vips.h>
    - C++ includes <vips/vipscpp.h>
- im_extract*() numbers bands from zero, not from 1 ... also other funcs which
  refer to bands (eg. im_lrmosaic() etc.)
- mosaic funcs have extra max blend width parameter

VIPS enhancements
=================

- rewritten and much fancier ip (see separate docs), now called nip ... old ip
  (slightly fixed) still there
- mmap windows remove 2GB image size limit
- images have an origin hint field
- icc transforms supported via wrapping of Little CMS
- much, much faster Fourier transforms via wrapping of libfwfft
- ppm/pgm/pbm read/write
- C++ API enhancements: new constructors, more operator equivalences
- new colourspace: Yxy
- support for D50 colour temperature images
- new functions: im_image(), im_system(), im_version*(), im_blend(),
  im_copy_set(), im_flood_blob(), im_icc_*(), im_open_local_array(),
  im_header_*(), im_sign(), im_ceil(), im_floor(), im_remainderconst*(),
  im_print(), im_guess_prefix(), im_remosaic(), im_invertlut(), Yxy funcs,
  color temperature funcs, im_clip2fmt(), im_blend(), im_lab_morph(),
  im_histnorm(), im_histcum(), im_video_v4l(), im_LabS2Lab(), im_Lab2LabS()
- new type: IMAGEVEC
- header is much faster
- ip/nip split to separate packages
- better vips2dj
- better cygwin support, mingw support too
- Mac OS X and Windows build support
- new set of sensible #define names, old names can be turned off with
  -DIM_NO_VIPS7_COMPAT
- many configure improvements

Bug fixes
=========

- speed ups
- fixes to implicit image format conversions 
- im_zoom() spots integer overflow
- im_LabS2LabQ() rounding on a/b slightly broken for a/b == 0
- fixes to refcounting in C++ API
- mask casts in C++ were broken




VIPS Changes for 7.4 to 7.6

Non-backwards compatible changes
================================

- im_histplot() has new (and better) rules
- im_andconst(), im_orconst() and im_eorconst() now take a
  double, not an unsigned char, as their constant argument type

libsrc/conversion/im_extract.c
/* @(#) Function to extract a portion of a Vasari format picture.
 * @(#) Function im_extract() assumes that the imin image
 * @(#) is either memory mapped or in the buffer pimin->data.
 * @(#)
 * @(#) int im_extract(pimin, pimout, pbox)
 * @(#) IMAGE *pimin, *pimout;
 * @(#) IMAGE_BOX *pbox;
 * @(#)
 * @(#) All functions return 0 on success and -1 on error
 * @(#)
 *
 * Copyright: 1990, J. Cupitt
 *
 * Author: J. Cupitt
 * Written on: 12/02/1990
 * Modified on: 4/6/92, J.Cupitt
 *	- speed up! why wasn't this done before? Why am I stupid?
 *	- layout, messages fixed
 * now extracts IM_CODING_LABQ to IM_CODING_LABQ file: K.Martinez 1/7/93
 * 2/7/93 JC
 *	- adapted for partial v2
 *	- ANSIfied
 * 7/7/93 JC
 *	- behaviour for IM_CODING_LABQ fixed
 *	- better messages
 * 7/10/94 JC
 *	- new IM_NEW()
 * 22/2/95 JC
 *	- new use of im_region_region()
 * 6/7/98 JC
 *	- im_extract_area() and im_extract_band() added
 * 11/7/01 JC
 *	- im_extract_band() now numbers from zero
 * 7/11/01 JC
 *	- oh what pain, im_extract now numbers bands from zero as well
 * 6/9/02 JC
 *	- zero xoff/yoff for extracted area
 * 14/4/04 JC
 *	- nope, -ve the origin
 * 17/7/04
 *	- added im_extract_bands(), remove many bands from image
 */

/*

    This file is part of VIPS.
    
    VIPS is free software; you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

/*

    These files are distributed with VIPS - http://www.vips.ecs.soton.ac.uk

 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif /*HAVE_CONFIG_H*/
#include <vips/intl.h>

#include <stdio.h>

#include <vips/vips.h>

#ifdef WITH_DMALLOC
#include <dmalloc.h>
#endif /*WITH_DMALLOC*/

/* Extract one or more bands ... get number of output bands from
 * or->im->Bands.
 */
static int
extract_band( REGION *or, REGION *ir, IMAGE *in, IMAGE_BOX *box )
{
	Rect iarea;
	Rect *r = &or->valid;
	int le = r->left;
	int ri = IM_RECT_RIGHT(r);
	int to = r->top;
	int bo = IM_RECT_BOTTOM(r);
	int es = IM_IMAGE_SIZEOF_ELEMENT( ir->im );	
	int ipel = IM_IMAGE_SIZEOF_PEL( ir->im );
	int opel = IM_IMAGE_SIZEOF_PEL( or->im );
	char *p, *q;
	int x, y, z;

	/* Ask for input we need.
	 */
	iarea = or->valid;
	iarea.left += box->xstart;
	iarea.top += box->ystart;
	if( im_prepare( ir, &iarea ) )
		return( -1 );

	for( y = to; y < bo; y++ ) {
		p = IM_REGION_ADDR( ir, le + box->xstart, y + box->ystart ) +
			box->chsel * es;
		q = IM_REGION_ADDR( or, le, y );

		for( x = le; x < ri; x++ ) {
			for( z = 0; z < opel; z++ )
				q[z] = p[z];

			p += ipel;
			q += opel;
		}
	}

	return( 0 );
}

/* Extract an area. Can just use pointers.
 */
static int
extract_area( REGION *or, REGION *ir, IMAGE *in, IMAGE_BOX *box )
{
	Rect iarea;

	/* Ask for input we need. Translate from demand in or's space to
	 * demand in ir's space.
	 */
	iarea = or->valid;
	iarea.left += box->xstart;
	iarea.top += box->ystart;
	if( im_prepare( ir, &iarea ) )
		return( -1 );

	/* Attach or to ir.
	 */
	if( im_region_region( or, ir, &or->valid, iarea.left, iarea.top ) )
		return( -1 );
	
	return( 0 );
}

/* Extract an area and bands from an image. 
 */
int
im_extract_areabands( IMAGE *in, IMAGE *out, 
	int left, int top, int width, int height, int band, int nbands )
{      
	IMAGE_BOX *box;

	if( im_piocheck( in, out ) )
		return( -1 );
        if( band < 0 || nbands < 1 || band + nbands > in->Bands ) {
                im_error( "im_extract_areabands", 
                        _( "band selection out of range" ) );
                return( -1 );
        }
	if( left + width > in->Xsize ||
		top + height > in->Ysize ||
		left < 0 || top < 0 ||
		width <= 0 || height <= 0 ) {
		im_error( "im_extract_areabands", _( "bad extract area" ) );
		return( -1 );
	}
        if( in->Coding != IM_CODING_NONE ) {
		if( in->Coding != IM_CODING_LABQ ) {
			im_error( "im_extract_areabands", 
				_( "unknown coding" ) );
			return( -1 );
		}

		/* We only do area extract for coding == labq.
		 */
		if( band != 0 || nbands != in->Bands ) {
			im_error( "im_extract_areabands", 
				_( "can only extract areas from LABQ" ) );
			return( -1 );
		}
        }

        /* Set up the output header.  
         */
        if( im_cp_desc( out, in ) ) 
                 return( -1 );
        out->Bands = nbands;
        out->Xsize = width;
        out->Ysize = height;
        if( im_demand_hint( out, IM_THINSTRIP, in, NULL ) )
                return( -1 );
        if( !(box = IM_NEW( out, IMAGE_BOX )) )
                return( -1 );
        box->xstart = left;
        box->ystart = top;
        box->xsize = width;
        box->ysize = height;
        box->chsel = band;
 
 	/* Extracting all bands is a special case ... we can do it with
	 * pointers.
	 */
	if( band == 0 && nbands == in->Bands ) {
		if( im_generate( out, 
			im_start_one, extract_area, im_stop_one, in, box ) )
			return( -1 );
	}
	else {
		if( im_generate( out, 
			im_start_one, extract_band, im_stop_one, in, box ) )
			return( -1 );
 	}
 
        out->Xoffset = -left;
        out->Yoffset = -top;

        return( 0 );
}

/* Legacy interface.
 */
int
im_extract( IMAGE *in, IMAGE *out, IMAGE_BOX *box )
{	
	if( box->chsel == -1 )
		return( im_extract_areabands( in, out, 
			box->xstart, box->ystart, box->xsize, box->ysize,
			0, in->Bands ) );
	else
		return( im_extract_areabands( in, out, 
			box->xstart, box->ystart, box->xsize, box->ysize,
			box->chsel, 1 ) );
}

/* Convenience functions.
 */
int
im_extract_area( IMAGE *in, IMAGE *out, 
	int left, int top, int width, int height )
{
	return( im_extract_areabands( in, out, 
		left, top, width, height, 0, in->Bands ) );
}

int
im_extract_bands( IMAGE *in, IMAGE *out, int chsel, int nbands )
{
	return( im_extract_areabands( in, out, 
		0, 0, in->Xsize, in->Ysize, chsel, nbands ) );
}

int
im_extract_band( IMAGE *in, IMAGE *out, int chsel )
{
	return( im_extract_bands( in, out, chsel, 1 ) ); 
}
